go/vt/vtgate/planbuilder/operators/apply_join.go

```diff
@@ -32,26 +32,50 @@ import (
 type ApplyJoin struct {
 	LHS, RHS ops.Operator
 
-	// Columns stores the column indexes of the columns coming from the left and right side
-	// negative value comes from LHS and positive from RHS
-	Columns []int
-
-	// ColumnsAST keeps track of what AST expression is represented in the Columns array
-	ColumnsAST []sqlparser.Expr
-
-	// Vars are the arguments that need to be copied from the LHS to the RHS
-	Vars map[string]int
-
 	// LeftJoin will be true in the case of an outer join
 	LeftJoin bool
 
 	// JoinCols are the columns from the LHS used for the join.
 	// These are the same columns pushed on the LHS that are now used in the Vars field
 	LHSColumns []*sqlparser.ColName
 
+	// Before offset planning
+
+	// ColumnsAST keeps track of what AST expression is represented in the Columns array
+	ColumnsAST []JoinColumn
+
+	// JoinPredicates are join predicates that have been broken up into left hand side and right hand side parts.
+	JoinPredicates []JoinColumn
+
+	// After offset planning
+
+	// Columns stores the column indexes of the columns coming from the left and right side
+	// negative value comes from LHS and positive from RHS
+	Columns []int
+
+	// Vars are the arguments that need to be copied from the LHS to the RHS
+	Vars map[string]int
+
 	Predicate sqlparser.Expr
 }
 
+// JoinColumn is where we store information about columns passing through the join operator
+// It can be in one of three possible configurations:
+//   - Pure left
+//     We are projecting a column that comes from the left. The RHSExpr will be nil for these
+//   - Pure right
+//     We are projecting a column that comes from the right. The LHSExprs will be empty for these
+//   - Mix of data from left and right
+//     Here we need to transmit columns from the LHS to the RHS,
+//     so they can be used for the result of this expression that is using data from both sides.
+//     All fields will be used for these
+type JoinColumn struct {
+	Original sqlparser.Expr // this is the original expression being passed through
+	BvNames  []string       // the BvNames and LHSCols line up
+	LHSExprs []sqlparser.Expr
```

Comment:
nit, I think you meant:

```suggestion
	BvNames  []string       // the BvNames and LHSExprs line up
	LHSExprs []sqlparser.Expr
```